proposals/break-continue-in-inline-lambdas.md

```diff
@@ -0,0 +1,141 @@
+to parameters without `noinline` or `crossinline` modifiers. 
+
+```kotlin
+@outer for (department in departments) {
+    for (employee in department.employees) {
+        employee.apply { // We're inside an inline lambda now
+            if (isRetired) continue // Unlabeled break/continue works with the closest enclosing for/while
+            if (position == "Developer" && age > 70) {
+                println("We have some senior developers in our organisation!")
+                break@outer
+            }
+        }
+    }
+}
+```
+
+## Motivation
+
+Inline functions with [trailing lambdas](https://kotlinlang.org/docs/lambdas.html#passing-trailing-lambdas) are designed in a way
+that allows to use them as if they were first-class language constructs (e.g. `synchronized` or `run`, `with`, `let` scope
+functions from stdlib).
+
+Since users can invoke `return`, `break` or `continue` inside of bodies of native language constructs (such as body of `if` or
+body of `try-catch`), it would be only logical to allow the same for lambdas passed as parameters to inline functions. At the
+moment only non-local `return`s are supported. That's why we are proposing to add support for non-local `break`/`continue` as
+well.
+
+So the Motivation is the following:
+
+1. Make inline lambdas truly first-class language constructs that are seamless to use.
+2. Fix consistency with non-local `return`. It's inconsistent that currently we allow non-local `return` statements but don't
+   allow non-local `break` and `continue` statements.
+
+## Design problems 
+
+This feature may interfere with another proposed [feature](https://youtrack.jetbrains.com/issue/KT-19748), 
+allowing to use `break`/`continue` not only inside loop statements, but also within lambda arguments
+of loop-like functions like `forEach` or `takeWhile`
+(the notion of a loop-like function would require a strict definition; we leave this question open for now).
+
+```kotlin
+for (i in 0 until 100) {
+    (0 until 200).forEach {
+        if (it == 42) break // Should it break from `for` or `forEach`?
+        if (i == 10) continue // Where to continue?
+    }
+}
+```
+
+However, it seems that both features can be implemented without introducing ambiguities.
+
+First of all, we should never allow unlabeled `break`/`continue` for "loop-like" functions because users
+wouldn't be able to understand whether it is applied to a certain function without checking its implementation.
+```kotlin
+foo { bar { break } } // which function is loop-like?
+```
+That leave us only with the labeled syntax option for "loop-like" functions:
+```kotlin
+for (i in 0 until 100) {
+    (0 until 200).forEach {
+        if (it == 42) break@forEach // Obviously, it breaks `forEach`
+        if (i == 10) continue // No label, so it applies to the 'for' loop
+    }
+}
+```
+
+A similar rule applies to `return` statements: an unlabeled return within a lambda is always non-local, a labeled one returns to the label.
```

Comment:
It is important to add here that an unlabeled `return` goes to an innermost enclosing block that is clearly marked with `fun` _hard_ keyword. In a similar way, an unlabeled `break`/`continue` go to an innermost enclosing block that is clearly marked with `for`, `do`, or `while` hard keywords.